### PR TITLE
[Merged by Bors] - feat: some results on dimension of field adjoin

### PR DESCRIPTION
**Pull request by acmepjz**
In `Mathlib/FieldTheory/Adjoin.lean`:

- `IntermediateField.rank_sup_le_of_isAlgebraic`: if `A` and `B` are intermediate fields, and at least one them are algebraic, then the rank of `A ⊔ B` is less than or equal to the product of that of `A` and `B`. (Note that this result is also true without algebraic assumption, but the proof becomes very complicated, and which is out of scope of this PR.)
- `IntermediateField.finrank_sup_le`: if `A` and `B` are intermediate fields, then the `Module.finrank` of `A ⊔ B` is less than or equal to the product of that of `A` and `B`.

Move the result `Subalgebra.[fin]rank_sup_le_of_free` from `Mathlib/LinearAlgebra/TensorProduct/Subalgebra.lean` to a new file `Mathlib/RingTheory/Adjoin/Dimension.lean`, as their proofs don't use tensor `mulMap` anymore.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/LinearAlgebra/TensorProduct/Field.lean
Comment (alreadydone) [inline Mathlib/LinearAlgebra/TensorProduct/Field.lean:31]: I think the transcendental case is not too complicated if you use item 1 of [my comment here](https://leanprover.zulipchat.com/#narrow/channel/217875-Is-there-code-for-X.3F/topic/Separability.20degree.3F/near/411585334). You've already proven the result for purely transcendental extensions [here](https://github.com/leanprover-community/mathlib4/pull/18741#issuecomment-2464787184). Even though we don't have a definition of transcendental degree yet, you can take a F-transcendental basis in A and one in B, and `A ⊔ B` will be algebraic over the union of the two bases, because of [IsTranscendenceBasis.isAlgebraic](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/AlgebraicIndependent.html#IsTranscendenceBasis.isAlgebraic) and [algebraicClosure](https://leanprover-community.github.io/mathlib4_docs/Mathlib/FieldTheory/AlgebraicClosure.html#algebraicClosure) is an intermediate field. (The union of the two bases needs not be a basis, but we should prove a general `IntermediateField.card(inal)_adjoin_le`.)
image

Comment (acmepjz) [reply Mathlib/LinearAlgebra/TensorProduct/Field.lean:31]: By "complicated" I mean the proof is lengthy, cumbersome and need a lot of imports. In fact I have a proof but there are sill some `sorry`s mainly for struggling the relation of 9 intermediate fields...

Maybe later I'll reconsider your comments and looking for a simpler proof. My opinion is that if we want to keep the import of this file minimal, then we can only prove this version. The general version should go to another file with heavier imports. What's your opinion?

Comment (acmepjz) [reply Mathlib/LinearAlgebra/TensorProduct/Field.lean:31]: > a general `IntermediateField.card(inal)_adjoin_le`

I think I have this result in my WIP file.

Comment (alreadydone) [reply Mathlib/LinearAlgebra/TensorProduct/Field.lean:31]: Is there a reason to keep imports minimal? This is a new file after all. By the way, how do you count import increases without opening a PR?

Comment (acmepjz) [reply Mathlib/LinearAlgebra/TensorProduct/Field.lean:31]: > Is there a reason to keep imports minimal? This is a new file after all.

No particular reason. But I feel that these two results are basic. I'm afraid that they will be used in other places. The direct use of them is in linearly disjointness, which does not import `AlgebraicIndependent` if I remembered correctly.

But anyway they imported `Algebraic`. The `AlgebraicIndependent` only adds 4 files compare to `Algebraic`. So maybe... it's reasonable.

> By the way, how do you count import increases without opening a PR?

I'm guessing from #18770 which imports `AlgebraicIndependent`.

Comment (acmepjz) [reply Mathlib/LinearAlgebra/TensorProduct/Field.lean:31]: I mean this https://github.com/leanprover-community/mathlib4/pull/18433#issuecomment-2446477384

Comment (alreadydone) [reply Mathlib/LinearAlgebra/TensorProduct/Field.lean:31]: Apparently it's showing the effect of adding two imports in Algebraic? Which leads to only one file increase in transitive imports of AlgebraicIndependent. Nothing there shows the effect of importing AlgebraicIndependent compared to importing Algebraic, or am I missing something?

Comment (acmepjz) [reply Mathlib/LinearAlgebra/TensorProduct/Field.lean:31]: I mean `Algebraic` imports 1314 files while `AlgebraicIndependent` imports 1317 files and which contains the previous 1314 files. Now the linearly disjointness file already imports `Algebraic`, so they import that 1314 files. If I want to import `AlgebraicIndependent`, then at most 3 new files need to be imported.

Comment (alreadydone) [reply Mathlib/LinearAlgebra/TensorProduct/Field.lean:31]: Indeed 4 of the 7 imports of AlgebraicIndependent is unnecessary; this looks like a bug of `shake` and I've reported [on Zulip](https://leanprover.zulipchat.com/#narrow/channel/287929-mathlib4/topic/import.20changes.20bot.20on.20PRs/near/481456065).

Comment (acmepjz) [reply Mathlib/LinearAlgebra/TensorProduct/Field.lean:31]: > Indeed 4 of the 7 imports of AlgebraicIndependent is unnecessary; this looks like a bug of `shake` and I've reported [on Zulip](https://leanprover.zulipchat.com/#narrow/channel/287929-mathlib4/topic/import.20changes.20bot.20on.20PRs/near/481456065).

It's probably because the file uses concepts defined in all of 7 files, but inside the 4 files they already import another 3 files.

> So the conclusion is that it's reasonable to import AlgebraicIndependent and prove the most general result right away, right?

It's reasonable to import `AlgebraicIndependent`. But for the proof there is still some way to go...

Comment (alreadydone) [reply Mathlib/LinearAlgebra/TensorProduct/Field.lean:31]: Oh, I realize that this result (algebraic case) should not actually require tensor products. I think you can take bases bA and bB of A and B over F, and show that `bA * bB` spans `A ⊔ B`, by using [Algebra.adjoin_eq_span_of_subset](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/Adjoin/Basic.html#Algebra.adjoin_eq_span_of_subset).

Comment (acmepjz) [reply Mathlib/LinearAlgebra/TensorProduct/Field.lean:31]: > Oh, I realize that this result (algebraic case) should not actually require tensor products. I think you can take bases bA and bB of A and B over F, and show that `bA * bB` spans `A ⊔ B`, by using ~[Algebra.adjoin_eq_span_of_subset](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/Adjoin/Basic.html#Algebra.adjoin_eq_span_of_subset)~ [Algebra.adjoin_union_coe_submodule](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/Adjoin/Basic.html#Algebra.adjoin_union_coe_submodule). (You'll need to prove a `Submodule.rank_mul_le` though, which should be easy from [Submodule.span_mul_span](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Algebra/Operations.html#Submodule.span_mul_span).

Interesting. Sounds that [Subalgebra.rank_sup_le_of_free](https://leanprover-community.github.io/mathlib4_docs/Mathlib/LinearAlgebra/TensorProduct/Subalgebra.html#Subalgebra.rank_sup_le_of_free) can also be proved without tensor products. I'll look at it.

Comment (acmepjz) [reply Mathlib/LinearAlgebra/TensorProduct/Field.lean:31]: `#find_home` on my computer just runs out of memory. @alreadydone Could you help me to find a suitable place to put this result?

Comment (alreadydone) [reply Mathlib/LinearAlgebra/TensorProduct/Field.lean:31]: Great! I usually use [the web editor](https://live.lean-lang.org/#codez=JYWwDg9gTgLgBAWQIYwBYBtgCMBQO0Cm0BIcAygK5ZLoDmBWUSAdEwHYDWA+gM4Vhd0BLhABmXUVAIEAhHLgBvAErk4ALjgAVAJ5gCAKgC+cANoBhCCBBLgbWnCUBdUwDkIbGFGAA3YDQeOOHDBphZWNnbkziYAgnQMTA5RQSEAFDHq5FQ09IxISWQAlHCpAEKZlNTxeQWFKcEmCBAAJhRCzABiUgRJMdFNre1d0kmlzmr1iC1tBKxInEmApYTpcIApRHClxYAmRFODs+wcvXD6uzNzCyrlagC8cFjaKRBYMEi2cIAX5ICdwDEANHcxgEvydS3AZnYazAgAD2APBgPC41B4MJKKhuDmKqQQwLgMTqwSeLzeX1KfywpSBaNBQ26zChMLhCKQSJ4KOxSgxWLRmxSqCQ3h6qGxcVyTGYSGaACsILYuBQ2MB3FwAMYQYR8LAgaZCJIrDRkAjwIolK7kA3kPFwKAAd1MMTFkulbC4BAAjrwwPNGcyklhfht7VKZa73Z7EciVGS/oAEwhx1VFfAEzQIoh+k0qmr2zB4HqdIDaIbYzhQcFQKSVvLsPSp+3mhxUyxgEHTWp6K3Wm22cC4KWtplQ0f+zBA3GDBwA5GOB2ShyO3eOx4FglJRLZWwcC4IelxCsxPPMWWYkFBmrYaDOuHn0Ju4AAfAB8u/YLIA/N3grQVXYoHAADwAbjvOAoSQJV4EPY9T3Qc92HoTc8AAYhXNhmi4VBLB6SocgSFh1wTTcRHESRpDkGQgA). In this case it finds Mathlib itself. Note that I replaced StrongRankCondition with Nontrivial but it doesn't affect the result. I'm not sure which current files can host this result without adding too many imports.

Comment (acmepjz) [reply Mathlib/LinearAlgebra/TensorProduct/Field.lean:31]: > I replaced StrongRankCondition with Nontrivial but it doesn't affect the result

Because in this case it uses an instance that a non-trivial commutative ring satisfies strong rank condition. I think that file is not imported by so many files.

> it finds Mathlib itself

Oops. What about `Mathlib.RingTheory.Adjoin.Dimension`?

Comment (alreadydone) [reply Mathlib/LinearAlgebra/TensorProduct/Field.lean:31]: I think for the Subalgebra result, Mathlib.LinearAlgebra.Dimension.Constructions is pretty reasonable. It has 2204 transitive imports and importing Mathlib.RingTheory.Adjoin.Basic adds 3 and importing Mathlib.Algebra.Group.Pointwise.Set.Card adds 1. ([web editor link](https://live.lean-lang.org/#codez=JYWwDg9gTgLgBAWQIYwBYBtgCMB0AZYAOwFMkoBBdAc2KyiRwBFRjCBnYCQnAYS7ZhQArgGMYndnAC0UuACY5ABgAscQUnbBxAN2JxQkWGwBQB6PGRpMuAEpEqAFVTFoATxzkAJgCsIRHABCSBwi0rIKigDsavSaOnpmRqbg5ogoGNge1LT0OADiUBBCYDgACn6EMADuwGzEOADKxDC8ZJ5h8koAHDEaHPH6KUnJhvB4pNwAouhIuHwgIBqexsYQYKxw4xr6hMYA9DJwAAYAxOrsAPqJMGxHcFDEo2xwqBBVcIuErr1xwLqDTzUzjgIiEUAelQ+EE8QnQelQwRwxlO5zYVyGNzgAigdyQ6DYECxryqzzQelRWj+xHQ32uxHaIGhsOIzyqrzqcEISBAeiwxCoRFZWlQx2xRyRUj2xmpszUAGsAFwAIjOsTR1zYSqxMAAHgqABTYgCUAH44AqQRAFks4ABeAB8cE8EGMcDgcPgGvNtrg+sACYRwGgwSaEbRGnB49AASQx9IQTLhADluSzXe7miCwVB4+0FT7gAAzdPwRmeAAC+pw2LgFxrRu9cADQeQRBzzKBGxAAEIlQByADepYAvr2XsE4FrqRylUq0x77iz4HmPigRCKBDq4DU0Gm3W6AD6crh6B0fHuzvd7w8EnnaqB2+27y/FgHPZcanAF4DoGDEbNIMA4ALIRCH0B8n2ffQi2AHAYAgBpBHsKsYDIG4AHVhTvHAgwQqB7A7UCbwSOApz0QhjwgvduyVAAdQg6P7Aw2CrYAAC9iCHbVUPw7cRX7bEhwVBimJwABHAlYF9AB22D4MQwgqDgAAeOAZLg3D7CNIcL3dCAqCjQgCwgch4BgOUzyVftQXBHMh0YjFmI4djOIpAYNX7B4BG04xVT6dFASVSwMlwWdjDJaBiBAOAGiELA8RoOgGFiOULjYYoLjhC4IALC4CweYgu0KuB+xsaLzTgBxXHWAAqTiAG15hAOwFLgGwAF04Dq3CuCoGwNDlPhCE8SkuFatqnwaq0mvwhoOrqygEvoVrovGy99XIcqYri7JEuWhp631AJNti+KciQPajQmts4RwAAxfLlvIObrvqe7iD0UqAg6hUnxenBkuWwBSwnWuBABSiOAAnrQATIkQBN6gB0qNuq2GYRuhGIYbLBXDTCAsBQog4EAC/JAE7gcgABo4CwchAEvyBs/re+piB1WobguOKOGefVSuXGwDoQBtyEut1cfx0CSYCCmsACWnl3p/KcCZlm0XZ2pfW5n1ed9fnl0htMEX+EVlwWs6Ix8CoLhAiQLhECBiFS2LS3bUqQYtJp4H230jtdjN9rTKB3nm03fCIC5iFE1KwA0NnglV0qqYpgIg/NsOI6jlXnjjiXGzgY3EqrdLPGIAsyafLbHZuthI8IC4QFhVPCA6lAXjTNcNBoFHmX+/qgf1dSHbh30NvByHoZrP2A9QCmAypnAQBSlPkt7Xsp8pxO59D8PF97VaFy/EhfWS+uMr0C5w1ROAeDaIg8VnlLa/QY+4H3e1YLVOATQuNMqBthT7yUgBuR0TMkBiAvlfLk6Bb4XFiDQY+KwZAnD3p4C4rxbxbVOnnQ+aUwDHyyjlPK71CpdiAA)) The script is from Damiano [here](https://leanprover.zulipchat.com/#narrow/channel/287929-mathlib4/topic/import.20changes.20bot.20on.20PRs/near/481467215).

File: Mathlib/FieldTheory/Adjoin.lean
Comment (alreadydone) [inline Mathlib/FieldTheory/Adjoin.lean:738]: LGTM! Let's merge this now; when you're able to remove the algebraic condition, it would be reasonable to import AlgebraicIndependent, which adds 3 imports, plus 1 for RelRank.

**Review by alreadydone — APPROVED** [review]
maintainer merge

File: Mathlib/FieldTheory/Adjoin.lean
Comment (acmepjz) [reply Mathlib/FieldTheory/Adjoin.lean:738]: > it would be reasonable to import AlgebraicIndependent, which adds 3 imports, plus 1 for RelRank

Unfortunately, that's not possible, since `AlgebraicIndependent` (`aevalEquivField` in my working PR) and `RelRank` imports `Mathlib.FieldTheory.Adjoin` 😂

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**
